### PR TITLE
Backport CMake build to v1.20

### PR DESCRIPTION
**Pull request by isaachier**

### CODE REVIEW

**Comment by isaachier** [comment]
Please change base branch to v1.20 when you can.

**Comment by ruslo** [comment]
> Please change base branch to v1.20 when you can.

You should have permissions for branch creation :)

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+)
+
+# POSIX code is specified separately so we can leave it out in the future.
+add_library(leveldb_port_posix OBJECT "")
+target_sources(leveldb_port_posix
+  PRIVATE
+    "${PROJECT_SOURCE_DIR}/port/port_posix.cc"
+    "${PROJECT_SOURCE_DIR}/port/port_posix_sse.cc"
+
+  PUBLIC
```
Comment (ruslo) [inline CMakeLists.txt]: What is the point of using `PRIVATE`/`PUBLIC` and `BUILD_INTERFACE` here? Why not just `add_library(... ${sources})`?

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+    "include/leveldb/slice.h"
+    "include/leveldb/status.h"
+    "include/leveldb/table_builder.h"
+    "include/leveldb/table.h"
+    "include/leveldb/write_batch.h"
+)
+target_include_directories(leveldb
+  PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
```
Comment (ruslo) [inline CMakeLists.txt]: `CMAKE_INSTALL_INCLUDEDIR` will be empty if you will not include https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+include(CheckIncludeFile)
+check_include_file("unistd.h" HAVE_UNISTD_H)
+
+include(CheckIncludeFileCXX)
+check_include_file_cxx("atomic" LEVELDB_ATOMIC_PRESENT)
+
+include(CheckLibraryExists)
+check_library_exists(crc32c crc32c_value "" HAVE_CRC32C)
+check_library_exists(snappy snappy_compress "" HAVE_SNAPPY)
```
Comment (ruslo) [inline CMakeLists.txt]: If it's 3rd party package then it should be managed by Hunter:
* https://github.com/ruslo/hunter/blob/62c52a53d5ec04192d087bef35ac6523abd2169c/examples/Snappy/CMakeLists.txt#L9-L13

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+test_big_endian(LEVELDB_IS_BIG_ENDIAN)
+
+include(CheckIncludeFile)
+check_include_file("unistd.h" HAVE_UNISTD_H)
+
+include(CheckIncludeFileCXX)
+check_include_file_cxx("atomic" LEVELDB_ATOMIC_PRESENT)
+
+include(CheckLibraryExists)
+check_library_exists(crc32c crc32c_value "" HAVE_CRC32C)
```
Comment (ruslo) [inline CMakeLists.txt]: 3rd party package? Should be added to Hunter first.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+      "${PROJECT_SOURCE_DIR}/util/testutil.h"
+  )
+  target_compile_definitions(leveldb_test_base
+    PUBLIC
+      LEVELDB_PLATFORM_POSIX
+  )
+  target_link_libraries(leveldb_test_base leveldb)
+
+  add_executable(autocompact_test
+    "${PROJECT_SOURCE_DIR}/db/autocompact_test.cc"
```
Comment (ruslo) [inline CMakeLists.txt]: Relative path will work fine here, you can use `db/autocompact_test.cc`.

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,470 @@
...
+      LEVELDB_PLATFORM_POSIX
+  )
+  target_link_libraries(leveldb_bench_base leveldb)
+
+  add_executable(db_bench
+    "${PROJECT_SOURCE_DIR}/db/db_bench.cc"
+  )
+  target_link_libraries(db_bench leveldb_bench_base)
+
+  check_library_exists(sqlite3 sqlite3_open "" HAVE_SQLITE3)
```
Comment (ruslo) [inline CMakeLists.txt]: If this code planned to be used while building with Hunter then should be managed by `hunter_add_package`.

File: CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+      COMPATIBILITY SameMajorVersion
+  )
+  install(
+    EXPORT LeveldbTargets
+    NAMESPACE Leveldb::
+    DESTINATION "${CMAKE_INSTALL_LIBDIR}/cmake/Leveldb"
+  )
+  install(
+    FILES
+      "${PROJECT_SOURCE_DIR}/LeveldbConfig.cmake"
```
Comment (ruslo) [inline CMakeLists.txt]: I don't see this file.

**Review by ruslo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (isaachier) [reply CMakeLists.txt]: I guess so. Will add this package.

File: CMakeLists.txt (hunk 3 above)
Comment (isaachier) [reply CMakeLists.txt]: Got it.

File: CMakeLists.txt (hunk 1 above)
Comment (isaachier) [reply CMakeLists.txt]: Not sure. This is pulled from https://github.com/pwnall/leveldb/tree/cmake, but I can change if you want.

File: CMakeLists.txt (hunk 2 above)
Comment (isaachier) [reply CMakeLists.txt]: Fair point. Will include that module.

File: CMakeLists.txt (hunk 5 above)
Comment (isaachier) [reply CMakeLists.txt]: Agreed

File: CMakeLists.txt (hunk 6 above)
Comment (isaachier) [reply CMakeLists.txt]: Right. Seeing as I think this is only for the benchmarking, I am not sure if better to fix or remove.

File: CMakeLists.txt (hunk 1 above)
Comment (ruslo) [reply CMakeLists.txt]: Up to you. I just think it's overcomplicated and introduce no any checks or functionality.

File: CMakeLists.txt (hunk 6 above)
Comment (ruslo) [reply CMakeLists.txt]: You can keep it, just check that `LEVELDB_BUILD_BENCHMARKS` is `OFF` by default or add `LEVELDB_BUILD_BENCHMARKS=OFF` explicitly:
* https://docs.hunter.sh/en/latest/creating-new/create/cmake.html#cmake-options

**Review by ruslo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,740 @@
+# Copyright 2017 The LEVELDB Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+cmake_minimum_required(VERSION 3.1)
+project(leveldb VERSION 0.1.0 LANGUAGES C CXX)
```
Comment (ruslo) [inline CMakeLists.txt]: `C CXX` is the default, no need to specify it explicitly. From [documentation](https://cmake.org/cmake/help/latest/command/project.html):
```
By default C and CXX are enabled if no language options are given.
```

File: CMakeLists.txt (hunk 4 above)
Comment (ruslo) [reply CMakeLists.txt]: Need `hunter_add_package`

File: CMakeLists.txt (hunk 3 above)
Comment (ruslo) [reply CMakeLists.txt]: Need `hunter_add_package`

File: CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,740 @@
...
+  "${PROJECT_BINARY_DIR}/include/port/port_config.h"
+)
+
+include_directories(
+  "${PROJECT_BINARY_DIR}/include"
+  "${PROJECT_SOURCE_DIR}"
+)
+
+# POSIX code is specified separately so we can leave it out in the future.
+<<<<<<< HEAD
```
Comment (ruslo) [inline CMakeLists.txt]: To remove

File: CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,740 @@
...
+    "include/leveldb/status.h"
+    "include/leveldb/table_builder.h"
+    "include/leveldb/table.h"
+    "include/leveldb/write_batch.h"
+>>>>>>> 4581febb2575667d4f6d377743c44e909ca0897c
+)
+target_include_directories(leveldb
+  PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+<<<<<<< HEAD
```
Comment (ruslo) [inline CMakeLists.txt]: To remove

**Review by ruslo — CHANGES_REQUESTED** [review]

File: CMakeLists.txt [hunk 11]
```diff
@@ -0,0 +1,363 @@
+# Copyright 2017 The LEVELDB Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file. See the AUTHORS file for names of contributors.
+
+cmake_minimum_required(VERSION 3.1)
+
+list(APPEND CMAKE_MODULE_PATH "${CMAKE_SOURCE_DIR}/cmake")
```
Comment (ruslo) [inline CMakeLists.txt]: same

File: CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+set(CMAKE_TOOLCHAIN_FILE "${CMAKE_SOURCE_DIR}/cmake/toolchain.cmake")
```
Comment (ruslo) [inline CMakeLists.txt]: same

File: CMakeLists.txt [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+)
+
+# TODO(pwnall): This is only needed for port_posix.
+set(THREADS_PREFER_PTHREAD_FLAG ON)
+find_package(Threads REQUIRED)
+target_link_libraries(leveldb Threads::Threads)
+set(deps Threads)
+
+if(LEVELDB_WITH_CRC32C)
+  target_link_libraries(leveldb crc32c::crc32c)
```
Comment (ruslo) [inline CMakeLists.txt]: same

File: cmake/toolchain.cmake [hunk 14]
```diff
@@ -0,0 +1,3 @@
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_STANDARD_REQUIRED OFF)
```
Comment (ruslo) [inline cmake/toolchain.cmake]: same

**Review by ruslo — CHANGES_REQUESTED** [review]

**Review by ruslo — APPROVED** [review]

**Comment by ruslo** [comment]
Released: https://github.com/hunter-packages/leveldb/releases/tag/v1.20-p0

**Pull request merged**